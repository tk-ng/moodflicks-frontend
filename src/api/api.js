import axios from "axios";

const baseURL = import.meta.env.VITE_BASE_URL || "http://localhost:3001/";

/** API Class.
 *
 * Static class tying together methods used to get/send to to the API.
 * There shouldn't be any frontend-specific stuff here, and there shouldn't
 * be any API-aware stuff elsewhere in the frontend.
 *
 */

export class Api {
	// the token for interactive with the API will be stored here.
	static token;

	static async request(url, data = {}, method = "get") {
		// console.debug("API Call:", url, data, method);

		const headers = { Authorization: `Bearer ${Api.token}` };
		const params = method === "get" ? data : {};

		try {
			return (
				await axios({
					url,
					method,
					baseURL,
					data,
					params,
					headers,
					allowAbsoluteUrls: false,
				})
			).data;
		} catch (err) {
			console.error("API Error:", err.response);
			let message = err.response.data.error.message;
			throw Array.isArray(message) ? message : [message];
		}
	}

	// Individual API routes

	static async getUser(username) {
		let res = await this.request(`/users/${username}`);
		return res.user;
	}

	static async saveChanges({ username, data }) {
		let res = await this.request(
			`/users/${username}`,
			{ username, data },
			"patch"
		);
		return res.user;
	}

	static async deleteUser(username) {
		let res = await this.request(`/users/${username}`, {}, "delete");
		return res.user;
	}

	static async login(data) {
		let res = await this.request("auth/token", data, "post");
		return res.token;
	}

	static async signup(data) {
		let res = await this.request("auth/register", data, "post");
		return res.token;
	}

	static async getTrendingMovies() {
		let res = await this.request("/movies");
		return res.movies;
	}

	static async searchMovie(keyword) {
		let res = await this.request(`/movies/search/?keyword=${keyword}`);
		return res.movies;
	}

	static async getMovieDetail(id) {
		let res = await this.request(`/movies/${id}`);
		return res.movie;
	}

	static async getMovieCredits(id) {
		let res = await this.request(`/movies/${id}/credits`);
		return res.credits;
	}

	static async getActivities() {
		let res = await this.request(`/moods/activities`);
		return res.activities;
	}

	static async searchMood(keyword) {
		let res = await this.request(`/moods/search/?keyword=${keyword}`);
		return res.moods;
	}

	static async getMoods() {
		let res = await this.request(`/moods`);
		return res.moods;
	}

	static async getMoodsByUser(username) {
		let res = await this.request(`/users/${username}/moods`);
		return res.moods;
	}

	static async getMood(id) {
		let res = await this.request(`/moods/${id}`);
		return res.mood;
	}

	static async getMoviesByMood(id) {
		let res = await this.request(`/moods/${id}/movies`);
		return res.movies;
	}

	static async getMoodsByMovie(id) {
		let res = await this.request(`/movies/${id}/moods`);
		return res.moods;
	}

	static async getSimilarMoods(id) {
		let res = await this.request(`/moods/${id}/similar`);
		return res.moods;
	}

	static async getRecommendedMoviesByMood(id) {
		let res = await this.request(`/moods/${id}/recommended`);
		return res.movies;
	}

	static async createMood({ mood, username }) {
		let res = await this.request(`/moods`, { mood, username }, "post");
		return res.mood;
	}

	// needs username to verify in backend
	static async saveMoodChanges({ id, name, username }) {
		let res = await this.request(`/moods/${id}`, { name, username }, "patch");
		return res.message;
	}

	// needs username to verify in backend
	static async addMovieToMood({ moodId, movieId, username }) {
		const movies = await this.getMoviesByMood(moodId);

		if (movies.length) {
			const movieIds = movies.map((m) => m.id);
			if (movieIds.includes(movieId)) return null;
		}

		let res = await this.request(
			`/moods/${moodId}/movies/${movieId}`,
			{ username },
			"post"
		);
		return res.movieMood;
	}

	// needs username to verify in backend
	static async removeMovieFromMood({ moodId, movieId, username }) {
		let res = await this.request(
			`/moods/${moodId}/movies/${movieId}`,
			{ username },
			"delete"
		);
		return res.message;
	}

	// needs username to verify in backend
	static async deleteMood({ id, username }) {
		let res = await this.request(`/moods/${id}`, { username }, "delete");
		return res.message;
	}

	static async voteMood({ moodId, voteType, username }) {
		let res = await this.request(
			`/moods/${moodId}/vote/${voteType}`,
			{ username },
			"post"
		);
		return res.message;
	}
}
