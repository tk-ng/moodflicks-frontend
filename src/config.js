const PosterBaseUrl = "https://image.tmdb.org/t/p/";

export { PosterBaseUrl };
