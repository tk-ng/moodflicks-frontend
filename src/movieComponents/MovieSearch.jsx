import { useState, useEffect } from "react";
import { useSearchParams } from "react-router";
import { Api } from "../api/api";
import { MovieCardList } from "./MovieCardList";
import { SearchBar } from "../SearchBar";

export const MovieSearch = () => {
	const [movies, setMovies] = useState([]);
	const [header, setHeader] = useState("Trending Movies");
	const [searchParams] = useSearchParams();
	const keyword = searchParams.get("keyword");

	async function getTrendingMovies() {
		try {
			const result = await Api.getTrendingMovies();
			setHeader("Trending Movies");
			setMovies(result);
		} catch (err) {
			console.error(err);
		}
	}

	async function movieSearch(keyword) {
		const movies = await Api.searchMovie(keyword);
		setHeader(`Search results for "${keyword}"`);
		setMovies(movies);
	}

	useEffect(() => {
		if (keyword && keyword.trim()) {
			movieSearch(keyword);
		} else {
			getTrendingMovies();
		}
	}, [keyword]);

	return (
		<>
			<SearchBar
				search={movieSearch}
				placeholder="Search for a movie title..."
			/>
			<h1 className="text-light">{header}</h1>
			{movies && movies.length ? (
				<MovieCardList movies={movies} />
			) : (
				<p>No movies found</p>
			)}
		</>
	);
};
