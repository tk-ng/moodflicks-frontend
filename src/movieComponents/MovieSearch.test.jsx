import { render, screen, waitFor } from "@testing-library/react";
import { MovieSearch } from "./MovieSearch";
import { MemoryRouter, useSearchParams } from "react-router";
import { Api } from "../api/api";
import { PosterBaseUrl } from "../config";

vi.mock("../api/api", () => ({
	Api: {
		getTrendingMovies: vi.fn(),
		searchMovie: vi.fn(),
	},
}));

vi.mock("react-router", async function () {
	const mod = await vi.importActual("react-router");
	return {
		...mod,
		useSearchParams: vi.fn(),
	};
});

describe("Render MovieSearch component without keyword search", function () {
	it("mounts trending movies without crashing (without movies)", async function () {
		useSearchParams.mockReturnValue([new Map()]);
		Api.getTrendingMovies.mockReturnValue([]);
		const { container, asFragment } = render(
			<MemoryRouter>
				<MovieSearch />
			</MemoryRouter>
		);
		expect(container).toBeInTheDocument();
		expect(asFragment).toMatchSnapshot();
		await waitFor(() =>
			expect(screen.getByText("No movies found")).toBeInTheDocument()
		);
	});

	it("mounts trending movies without crashing (with movies)", async function () {
		useSearchParams.mockReturnValue([new Map()]);
		Api.getTrendingMovies.mockReturnValue([
			{
				id: "testid1",
				posterPath: "/testposter1.jpg",
				title: "Test Movie 1",
				releaseDate: "2025-02-28",
				overview: "Test Movie 1 Overview.",
			},
			{
				id: "testid2",
				posterPath: "/testposter2.jpg",
				title: "Test Movie 2",
				releaseDate: "2025-02-28",
				overview: "Test Movie 2 Overview.",
			},
		]);
		render(
			<MemoryRouter>
				<MovieSearch />
			</MemoryRouter>
		);
		await waitFor(() =>
			expect(
				screen.getByText("Test Movie 1", { selector: "h5" })
			).toBeInTheDocument()
		);
		screen
			.getAllByAltText("Movie Poster")
			.forEach((el, idx) =>
				expect(el).toHaveAttribute(
					"src",
					PosterBaseUrl + "w185" + `/testposter${idx + 1}.jpg`
				)
			);
	});
});
describe("Render MovieSearch component with keyword search", function () {
	it("mounts trending movies without crashing (without movies)", async function () {
		useSearchParams.mockReturnValue([
			new Map().set("keyword", "testkeywordnoresult"),
		]);
		Api.searchMovie.mockReturnValue([]);
		const { container, asFragment } = render(
			<MemoryRouter>
				<MovieSearch />
			</MemoryRouter>
		);
		expect(container).toBeInTheDocument();
		expect(asFragment).toMatchSnapshot();
		await waitFor(() =>
			expect(screen.getByText("No movies found")).toBeInTheDocument()
		);
	});

	it("mounts trending movies without crashing (with movies)", async function () {
		useSearchParams.mockReturnValue([new Map().set("keyword", "Test Movie")]);
		Api.searchMovie.mockReturnValue([
			{
				id: "testid1",
				posterPath: "/testposter1.jpg",
				title: "Test Movie 1",
				releaseDate: "2025-02-28",
				overview: "Test Movie 1 Overview.",
			},
			{
				id: "testid2",
				posterPath: "/testposter2.jpg",
				title: "Test Movie 2",
				releaseDate: "2025-02-28",
				overview: "Test Movie 2 Overview.",
			},
		]);
		render(
			<MemoryRouter>
				<MovieSearch />
			</MemoryRouter>
		);
		await waitFor(() =>
			expect(
				screen.getByText("Test Movie 1", { selector: "h5" })
			).toBeInTheDocument()
		);
		screen
			.getAllByAltText("Movie Poster")
			.forEach((el, idx) =>
				expect(el).toHaveAttribute(
					"src",
					PosterBaseUrl + "w185" + `/testposter${idx + 1}.jpg`
				)
			);
	});
});
