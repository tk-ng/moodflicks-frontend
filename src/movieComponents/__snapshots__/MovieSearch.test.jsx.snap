// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Render MovieSearch component with keyword search > mounts trending movies without crashing (without movies) 1`] = `[Function]`;

exports[`Render MovieSearch component without keyword search > mounts trending movies without crashing (without movies) 1`] = `[Function]`;

exports[`Render MovieSearch component without keyword search > mounts without crashing without movies 1`] = `[Function]`;
